overlord/configstate/transaction.go

```diff
@@ -114,7 +152,106 @@ func (t *Transaction) GetMaybe(snapName, key string, result interface{}) error {
 	return nil
 }
 
-// Commit saves to the state the configuration changes made in the transaction.
+func getFromPristine(snapName string, subkeys []string, pos int, config map[string]*json.RawMessage, result interface{}) error {
+	raw, ok := config[subkeys[pos]]
+	if !ok {
+		return &NoOptionError{SnapName: snapName, Key: strings.Join(subkeys[:pos+1], ".")}
+	}
+
+	if pos+1 == len(subkeys) {
+		err := json.Unmarshal([]byte(*raw), result)
+		if err != nil {
+			key := strings.Join(subkeys, ".")
+			return fmt.Errorf("internal error: cannot unmarshal snap %q option %q into %T: %s, json: %s", snapName, key, result, err, *raw)
+		}
+		return nil
+	}
+
+	var configm map[string]*json.RawMessage
+	err := json.Unmarshal([]byte(*raw), &configm)
+	if err != nil {
+		return fmt.Errorf("snap %q option %q is not a map", snapName, strings.Join(subkeys[:pos+1], "."))
+	}
+	return getFromPristine(snapName, subkeys, pos+1, configm, result)
+}
+
+func getFromChange(snapName string, subkeys []string, pos int, config map[string]interface{}, result interface{}) error {
+	value, ok := config[subkeys[pos]]
+	if !ok {
+		return &NoOptionError{SnapName: snapName, Key: strings.Join(subkeys[:pos+1], ".")}
+	}
+
+	if pos+1 == len(subkeys) {
+		raw, ok := value.(*json.RawMessage)
+		if !ok {
+			raw = jsonRaw(value)
+		}
+		err := json.Unmarshal([]byte(*raw), result)
+		if err != nil {
+			key := strings.Join(subkeys, ".")
+			return fmt.Errorf("internal error: cannot unmarshal snap %q option %q into %T: %s, json: %s", snapName, key, result, err, *raw)
+		}
+		return nil
+	}
+
+	configm, ok := value.(map[string]interface{})
+	if !ok {
+		raw, ok := value.(*json.RawMessage)
+		if !ok {
+			raw = jsonRaw(value)
+		}
+		err := json.Unmarshal([]byte(*raw), &configm)
+		if err != nil {
+			return fmt.Errorf("snap %q option %q is not a map", snapName, strings.Join(subkeys[:pos+1], "."))
+		}
+	}
+	return getFromChange(snapName, subkeys, pos+1, configm, result)
+}
+
+func patchConfig(snapName string, subkeys []string, pos int, config interface{}, value *json.RawMessage) (interface{}, error) {
+
+	switch config := config.(type) {
+	case nil:
+		// Missing update map. Create and nest final value under it.
+		configm := make(map[string]interface{})
+		_, err := patchConfig(snapName, subkeys, pos, configm, value)
+		if err != nil {
+			return nil, err
+		}
+		return configm, nil
+
+	case *json.RawMessage:
```

Comment:
It should happen on any final value. Anything set=<value>, value is marshaled before going into the tree.

Per the logic below (and tests), it breaks politely if the value is not a map though.